relayer/workers/beefyrelayer/beefy-ethereum-listener.go

```diff
@@ -153,26 +151,88 @@ func (li *BeefyEthereumListener) queryLightClientEvents(ctx context.Context, sta
 	return events, nil
 }
 
-// processLightClientEvents matches events to BEEFY commitment info by transaction hash
-func (li *BeefyEthereumListener) processLightClientEvents(ctx context.Context, events []*lightclientbridge.ContractInitialVerificationSuccessful) {
+// processHistoricalInitialVerificationSuccessfulEvents processes historical InitialVerificationSuccessful
+// events, updating the status of matched BEEFY justifications in the database
+func (li *BeefyEthereumListener) processHistoricalInitialVerificationSuccessfulEvents(ctx context.Context,
+	blockNumber, latestBlockNumber uint64) {
+
+	// Query previous InitialVerificationSuccessful events and update the status of BEEFY justifications in database
+	events, err := li.queryInitialVerificationSuccessfulEvents(ctx, blockNumber, &latestBlockNumber)
+	if err != nil {
+		li.log.WithError(err).Error("Failure fetching event logs")
+	}
+
+	li.log.Info(fmt.Sprintf(
+		"Found %d InitialVerificationSuccessful events between blocks %d-%d",
+		len(events), blockNumber, latestBlockNumber),
+	)
+
 	for _, event := range events {
-		// Only process events emitted by transactions sent from our node
-		if event.Prover != li.ethereumConn.GetKP().CommonAddress() {
-			continue
+		// Fetch validation data from contract using event.ID
+		validationData, err := li.lightClientBridge.ContractCaller.ValidationData(nil, event.Id)
+		if err != nil {
+			li.log.WithError(err).Error(fmt.Sprintf("Error querying validation data for ID %d", event.Id))
+		}
+
+		// Attempt to match items in database based on their payload
+		itemFoundInDatabase := false
+		items := li.beefyDB.GetItemsByStatus(store.CommitmentWitnessed)
+		for _, item := range items {
```

Comment:
Can we not query commitments directly, e. g. by hash? This could get super slow, depending on how many blocks we need to catch up on and how many commitments are in the database and how many blocks we need to catch up on.

Reply:
Which hash are you referring to here?

Reply:
This current logic is getting all witnessed commitments from the DB, and then we iterate over them. If there are millions of commitments in the DB, this will be super expensive. Could we instead query the commitment directly by ID/hash/index?

Reply:
- Data from the BEEFY commitment: `ValidatorAddresses`, `SignedCommitment` (we can generate the `Payload` from this)
- Data from the initial verification tx: on-contract commitment ID (we can use this to query `Payload` from contract)
- Data from the second verification tx: on-contract commitment ID (and therefore `Payload`)

In live mode we're doing these lookups using hashes from txs we send, but in sync mode in order to associate a BEEFY commitment with an initial verification tx we need to compare the `Payload` because we don't have any other information available. We could adopt the contract's commitment ID and use it to refer to BEEFY commitments in the relayer database, but even then we'd need to compare the `Payload` to make the initial association.

Maybe in sync mode we could subscribe to our own address' sent transactions so that they're available to make these lookups? Open to ideas on this one.

Reply:
@philipstanislaus given the scope of these (potential) changes are you okay if I make an issue for this so we can get the functional current version merged into master?

Reply:
Yes, definitely 👍

Reply:
Created issue https://github.com/Snowfork/polkadot-ethereum/issues/414